### PR TITLE
Add a module callback to react to account data changes

### PR DESCRIPTION
**Pull request by babolivier**
Part of the Tchap mainlining.

This is in order to implement the new plan for user discovery. See https://github.com/matrix-org/synapse-dinsic/issues/125 for the high-level description and https://docs.google.com/document/d/1wQw1OMgqn0blbv4Jt7ApZhR8mtN39ZLkNHKuNY-Xd_M/edit for the detailed action plan (which this work is a part of).

More specifically, this callback will be used by the module that will maintain the "red list" to change a user's visibility when the right account data changes.

### CODE REVIEW

**Review by anoadragon453 — COMMENTED** [review]
Drive-by comments on a couple small things as I saw this PR go past.

File: docs/modules/account_data_callbacks.md [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+## Callbacks
+
+The available presence router callbacks are:
+
+### `on_account_data_change`
+
+_First introduced in Synapse v1.57.0_
+
+```python
+async def on_account_data_change(
```
[the remark this answers is not in the document]
Comment (squahtx) [reply docs/modules/account_data_callbacks.md:16]: Looks like we also have an `on_user_deactivation_status_changed` callback.

File: docs/modules/account_data_callbacks.md [hunk 2]
```diff
@@ -0,0 +1,106 @@
+# Account data callbacks
+
+Account data callbacks allow module developers to react to change with the account data
```
Comment (squahtx) [inline docs/modules/account_data_callbacks.md:3]:
```suggestion
Account data callbacks allow module developers to react to changes of the account data
```

File: docs/modules/account_data_callbacks.md [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+of local users. Account data callbacks can be registered using the module API's
+`register_account_data_callbacks` method.
+
+## Callbacks
+
+The available presence router callbacks are:
```
Comment (squahtx) [inline docs/modules/account_data_callbacks.md:9]:
```suggestion
The available account data callbacks are:
```

File: docs/modules/account_data_callbacks.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (babolivier) [reply docs/modules/account_data_callbacks.md:16]: Agreed, let's go with past tense and try to stick to it.

**Review by squahtx — APPROVED** [review]
lgtm

I'm happy if @anoadragon453 is happy

**Comment by babolivier** [comment]
Andrew said his were just some drive-by comments so since they got addressed I'm going to assume he's also happy and merge this.

**Pull request merged**